src/arithmetic/algebraic_expression/algebraic_expression_optimization.c

```diff
@@ -0,0 +1,200 @@
+				C = r;
+
+				AlgebraicExpression_Free(l);
+				AlgebraicExpression_AddChild(lMul, A);
+				AlgebraicExpression_AddChild(lMul, C);
+				AlgebraicExpression_AddChild(rMul, B);
+				AlgebraicExpression_AddChild(rMul, AlgebraicExpression_Clone(C));
+
+				// TODO: Mark r as reusable.
+				// if(r->type == AL_OPERATION) r->operation.reusable = true;
+			} else {
+				// Righthand side is addition.
+				// C * (A + B) = (C * A) + (C * B)
+
+				A = _AlgebraicExpression_OperationRemoveLeftmostChild(r);
+				B = _AlgebraicExpression_OperationRemoveRightmostChild(r);
+				C = l;
+
+				AlgebraicExpression_Free(r);
+				AlgebraicExpression_AddChild(lMul, C);
+				AlgebraicExpression_AddChild(lMul, A);
+				AlgebraicExpression_AddChild(rMul, AlgebraicExpression_Clone(C));
+				AlgebraicExpression_AddChild(rMul, B);
+
+				// TODO: Mark r as reusable.
+				// if(l->type == AL_OPERATION) l->operation.reusable = true;
+			}
+			*root = add;
+			return true;
+		}
+	}
+
+	// Recurse.
+	uint child_count = AlgebraicExpression_ChildCount(*root);
+	for(uint i = 0; i < child_count; i++) {
+		if(__AlgebraicExpression_MulOverSum((*root)->operation.children + i)) return true;
+	}
+	return false;
+}
+
+// Distributive, multiplication over addition:
+// A * (B + C) = (A * B) + (A * C)
+//
+//           (*)
+//   (A)             (+)
+//            (B)          (C)
+//
+// Becomes
+//
+//               (+)
+//       (*)                (*)
+// (A)        (B)     (A)        (C)
+//
+// Whenever we encounter a multiplication operation
+// where one child is an addition operation and the other child
+// is a multiplication operation, we'll replace root multiplication
+// operation with an addition operation with two multiplication operations
+// one for each child of the original addition operation, as can be seen above.
+// we'll want to reuse the left handside of the multiplication.
+static void _AlgebraicExpression_MulOverSum(AlgebraicExpression **root) {
+	// As long as the tree changes keep modifying.
+	while(__AlgebraicExpression_MulOverSum(root));
+}
+
+/* Collapse multiplication operation under a single multiplication op
+ * exp = A * B * C
+ * exp can be computed in two ways:
+ * 1. (A * B) * C
+ * 2. A * (B * C)
+ * by flattening the expression both 1 and 2 are represented by the
+ * same tree structure. */
+static void _AlgebraicExpression_FlattenMultiplications(AlgebraicExpression *root) {
+	assert(root);
+	uint child_count;
+
+	switch(root->type) {
+	case AL_OPERATION:
+		switch(root->operation.op) {
+		case AL_EXP_ADD:
+		case AL_EXP_TRANSPOSE:
+			// Keep searching for a multiplication operation.
+			child_count = AlgebraicExpression_ChildCount(root);
+			for(int i = 0; i < child_count; i++) {
+				_AlgebraicExpression_FlattenMultiplications(CHILD_AT(root, i));
+			}
+			break;
+
+		case AL_EXP_MUL:
+			// Root has sub multiplication node(s).
+			if(AlgebraicExpression_OperationCount(root, AL_EXP_MUL) > 1) {
```

Comment:
Good catch, I don't see a way now avoiding performing `MulOverSum` for cases where both
direct children are addition operation, e.g. (A+B)*(C+D) will be transformed to (A*C)+(A*D)+(B*C)+(B*D) this will be done go guarantee the final tree structure, a multiplication operation will have either operands or transpose as its children.

Reply:
An example query: `MATCH ()-[:A|B]->()-[:C|D]->()`